Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs

```diff
@@ -0,0 +1,196 @@
+
+        if (!container.Solutions.TryGetValue(name, out var solutionSlot))
+        {
+            solutionSlot = ContainerSystem.EnsureContainer<ContainerSlot>(entity.Owner, $"solution@{name}");
+            container.Solutions.Add(name, solutionSlot);
+            Dirty(uid, container);
+        }
+
+        var needsInit = false;
+        SolutionComponent solutionComp;
+        if (solutionSlot.ContainedEntity is not { } solutionId)
+        {
+            (solutionId, solutionComp) = SpawnSolutionUninitialized(uid, name, new Solution() { MaxVolume = minVol, Name = name });
+            ContainerSystem.Insert(solutionId, solutionSlot, force: true);
+            existed = false;
+            needsInit = true;
+        }
+        else if (!TryComp(solutionId, out solutionComp!))
+        {
+            solutionComp = new SolutionComponent()
+            {
+                Solution = new Solution() { MaxVolume = minVol, Name = name },
+            };
+            AddComp(solutionId, solutionComp);
+            existed = false;
+        }
+        else
+        {
+            solutionComp.Solution.MaxVolume = FixedPoint2.Max(solutionComp.Solution.MaxVolume, minVol);
+            Dirty(solutionId, solutionComp);
+        }
+
+        if (!TryComp(solutionId, out ContainerSolutionComponent? relation))
+        {
+            relation = new ContainerSolutionComponent() { Container = uid, Name = name };
+            AddComp(solutionId, relation);
+        }
+        else
+        {
+            (relation.Container, relation.Name) = (uid, name);
+            Dirty(solutionId, relation);
+        }
+
+        if (needsInit)
+            EntityManager.InitializeAndStartEntity(solutionId, Transform(solutionId).MapID);
+
+        return (solutionId, solutionComp);
+    }
+
+    private Solution EnsureSolutionPrototype(Entity<SolutionContainerManagerComponent?> entity, string name, FixedPoint2 minVol, out bool existed)
+    {
+        existed = true;
+
+        var (uid, container) = entity;
+        if (!Resolve(uid, ref container, logMissing: false))
+        {
+            container = AddComp<SolutionContainerManagerComponent>(uid);
+            existed = false;
+        }
+
+        if (!container.Solutions.TryGetValue(name, out var prototype))
+        {
+            prototype = new Solution() { MaxVolume = minVol, Name = name };
+            container.Solutions.Add(name, prototype);
+            existed = false;
+        }
+        else
+            prototype.MaxVolume = FixedPoint2.Max(prototype.MaxVolume, minVol);
+
+        return prototype;
+    }
+
+
+    private Entity<SolutionComponent> SpawnSolutionUninitialized(EntityUid container, string name, Solution prototype)
+    {
+        var coords = new EntityCoordinates(container, Vector2.Zero);
+        var uid = EntityManager.CreateEntityUninitialized(null, coords, null);
+        if (!TryComp(uid, out SolutionComponent? comp))
+        {
+            var solution = prototype.Clone();
```

Comment:
This is currently only ever called with a new solution, so this is currently always an unnecessary clone.
Seeing as this is a private method anyways, it should be fine to just use the solution passed in without any cloning.

Reply:
Done.